### ISSUE TITLE
[bug] use-memo rule gives a false error when using a typescript cast

### ISSUE THREAD

**Issue by lukpsaxo** [Rel1cx/eslint-react#1583]
### Describe the bug

Thanks for this project btw! We are looking at migrating a very large codebase which is where the bugs come from.

If you add `as X` to do a typescript cast after the useMemo you can get an error that you are not assigning the result to a variable

### Reproduction

```
function A() {
    const x = useMemo(() => {
        return 1 + 1;
    }, []) as number;
    return x;
}
```

The error disappears if you remove `as number` and appears again if you add it back

### Expected behavior

No error

### Platform and versions

```Text
windows and 3.0.0-beta.60
```

### Stack trace

```Text

```

**Comment by lukpsaxo** [comment]
btw easy fix for us is moving the cast inside the memo.
